browser/chat/chat_tab_helper.cc

```diff
@@ -0,0 +1,207 @@
+
+std::string ChatTabHelper::GetConversationHistoryAsString() {
+  std::string history_text;
+
+  for (ConversationTurn history : chat_history_) {
+    history_text = base::StrCat({history_text,
+                                 history.character_type == CharacterType::HUMAN
+                                     ? chat_ui::kHumanPrompt
+                                     : chat_ui::kAIPrompt,
+                                 history.text, "\n"});
+  }
+
+  return history_text;
+}
+
+std::vector<ConversationTurn> ChatTabHelper::GetConversationHistory() {
+  return chat_history_;
+}
+
+void ChatTabHelper::AddToConversationHistory(const ConversationTurn& turn) {
+  chat_history_.push_back(turn);
+}
+
+void ChatTabHelper::GetArticleSummaryString(
+    OnArticleSummaryCallback on_article_summary_callback) {
+  if (!article_summary_.empty()) {
+    VLOG(1) << __func__ << " Summary coming from cache: " << article_summary_;
+
+    std::move(on_article_summary_callback).Run(article_summary_, true);
+    return;
+  }
+
+  ui::AXTreeID tree_id = web_contents()->GetPrimaryMainFrame()->GetAXTreeID();
+  content::RenderFrameHost* rfh =
+      content::RenderFrameHost::FromAXTreeID(tree_id);
+
+  rfh->GetMainFrame()->RequestAXTreeSnapshot(
+      base::BindOnce(&ChatTabHelper::OnSnapshotFinished, base::Unretained(this),
+                     std::move(on_article_summary_callback)),
+      ui::AXMode::kWebContents, false, 5000, {});
+}
+
+void ChatTabHelper::SetArticleSummaryString(const std::u16string text) {
+  article_summary_ = text;
+}
+
+void ChatTabHelper::AddObserver(Observer* observer) {
+  observers_.AddObserver(observer);
+}
+
+void ChatTabHelper::RemoveObserver(Observer* observer) {
+  observers_.RemoveObserver(observer);
+}
+
+void ChatTabHelper::OnSnapshotFinished(
+    OnArticleSummaryCallback on_article_summary_callback,
+    const ui::AXTreeUpdate& snapshot) {
+  ui::AXTree tree;
+  if (!tree.Unserialize(snapshot)) {
+    return;
+  }
+
+  // Start AX distillation process
+  // Don't copy the tree, as it can be expensive.
+  DistillViaAlgorithm(std::move(on_article_summary_callback), std::move(tree));
+}
+
+void ChatTabHelper::DistillViaAlgorithm(
+    OnArticleSummaryCallback&& on_article_summary_callback,
+    const ui::AXTree&& tree) {
+  std::vector<const ui::AXNode*> content_root_nodes;
+  std::vector<const ui::AXNode*> content_nodes;
+  GetContentRootNodes(tree.root(), &content_root_nodes);
+
+  for (const ui::AXNode* content_root_node : content_root_nodes) {
+    AddContentNodesToVector(content_root_node, &content_nodes);
+  }
+
+  std::vector<std::u16string> text_node_contents;
+  for (const ui::AXNode* content_node : content_nodes) {
+    AddTextNodesToVector(content_node, &text_node_contents);
+  }
+
+  // TODO(nullhook): The assumption here is that 9300 chars equate to
+  // approximately 2k tokens, which is a rough estimate. A proper tokenizer is
+  // needed for accurate measurement.
+  auto contents_text =
```

Comment:
Looks like chromium's tflite helper has classes to help with tokenization. Perhaps, look into this later: https://source.chromium.org/chromium/chromium/src/+/main:third_party/tflite_support/src/tensorflow_lite_support/cc/text/tokenizers/sentencepiece_tokenizer.h